src/Common/SettingConstraintType.h

```diff
@@ -0,0 +1,24 @@
+#pragma once
+
+
+namespace DB
+{
+
+enum class SettingConstraintType
+{
+    // Default. Behave in the same way as WRITABLE, but is not inherited unless `settings_constraints_replace_previous` is set.
+    NONE,
+
+    // Setting can be change within specified range only in `readonly=0` or `readonly=2` mode.
+    WRITABLE,
```

Comment:
Shouldn't be this the default mode?

Reply:
Currently writability is described by `optional<bool> readonly` mostly. And the mapping I came up with:
1. optional is not set - NONE (I guess it's better to rename to DEFAULT to be more precise)
2. set to false - WRITABLE; it can be explicitly set from SQL and, if I get it correctly its inheritance is different (see comments).
3. set to true - RADONLY.
imho It would be better to get rid of (1) at all, but it just looked too scary and big to rewrite. Maybe do it in a separate PR?

Reply:
As far as I remember `optional<bool>` was used to represent the situation when READONLY or WRITABLE are not written in ALTER command. Actually I'm not sure it was worth it.

Reply:
It seems now if we define any constraint it's always has its writability defined even if this is a command not specifying it explicitly, like `ALTER USER ... SETTINGS a=10 MIN 5`

Reply:
`std::optional<SettingsConstraintsWritability>` seems better then because it can be unset but if it's set it can have one of three values

Reply:
Done